### PR TITLE
[OpenMP][OMPIRBuilder] Error propagation across callbacks

### PR DESCRIPTION
**Pull request by skatrak**
This patch implements an approach to communicate errors between the OMPIRBuilder and its users. It introduces `llvm::Error` and `llvm::Expected` objects to replace the values returned by callbacks passed to `OMPIRBuilder` codegen functions. These functions then check the result for errors when callbacks are called and forward them back to the caller, which has the flexibility to recover, exit cleanly or dump a stack trace.

This prevents a failed callback to leave the IR in an invalid state and still continue the codegen process, triggering unrelated assertions or segmentation faults. In the case of MLIR to LLVM IR translation of the 'omp' dialect, this change results in the compiler emitting errors and exiting early instead of triggering a crash for not-yet-implemented errors. The behavior in Clang and openmp-opt stays unchanged, since callbacks will continue always returning 'success'.

### CODE REVIEW

**Comment by Meinersbur** [comment]
My naive view would be that once we start emitting code, other than non-recoverable errors such as bugs and non-yet-implemented aborts where the compiler should just crash. For descriptive messages for crashes, there is https://github.com/llvm/llvm-project/blob/df0551298868b164197a4e54e9444120dc96ff53/llvm/include/llvm/Support/PrettyStackTrace.h#L52 . This is what adds
```
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/clang++ -gdwarf-4 -g -o /app/output.s -mllvm --x86-asm-syntax=intel -fno-verbose-asm -S --gcc-toolchain=/opt/compiler-explorer/gcc-snapshot -fcolor-diagnostics -fno-crash-diagnostics <source>
1.	<eof> parser at end of file
2.	<source>:2:6: LLVM IR generation of declaration 'test'
3.	<source>:2:6: Generating code for declaration 'test'
```
to Clang crash reports. "failed region translation" (or more descriptive: "translating region 'X'") could be an entry added in OpenMPIRBuilder.

In other words, what recoverable error will this be used for? `convertBlock` not succeeding looks non-recoverable.

**Comment by Meinersbur** [comment]
I think continuing on an unrecoverable error is an inherently bad idea. It will create code paths that will never be tested by definition (because if there is a path we will fix the bug leading to that path) and hence likely cause dereferenced null pointers, trigger assertions, etc already because the LLVM generated so far is not in an expected state. For instance, the new early abort in `createTaskgroup` may leave a degenerate BB, any call to `BB::getTerminator()` will return nullptr, and at latest the IR-verifier will crash.  Any `llvm_unreachable` and `assert` in OpenMPIRBuilder and elsewhere will also still cause a crash, not `mlir::emitError` which seems inconsequential.

However, of course I will not stand in the way if this is the community-chosen solution and/or established pattern. Also, if there is use case with a recoverable error we would need this as well.

**Comment by tblah** [comment]
I don't feel strongly about this, but I think the proposed solution looks like a step forward. I think it is important that we do error out as soon as possible after an unrecoverable error and I think this is a cleaner way to do that than setting captured variables like `bodyGenStatus`.

I think the value in having some controlled error reporting like this over an assertion is for cases where implementation is still in progress. This allows us to print something more user friendly than a segfault/stack trace. As a reviewer I would read something like these errors as "this is a known limitation or error condition" whereas an assertion as "this is assumed never to happen".

**Comment by Meinersbur** [comment]
I might have overblown the problem. If with every `Error`/`Expected` we immediately return from the entire call stack, until there is no use of the LLVM-IR, there should not be a lot of things that could go wrong. Clang's use of OpenMPIRBuilder still needs to crash on an erroneous OpenMPIRBuilder call because it doesn't have that `Error`/`Expected` design.

**Comment by skatrak** [comment]
> I think continuing on an unrecoverable error is an inherently bad idea. It will create code paths that will never be tested by definition (because if there is a path we will fix the bug leading to that path) and hence likely cause dereferenced null pointers, trigger assertions, etc already because the LLVM generated so far is not in an expected state. For instance, the new early abort in `createTaskgroup` may leave a degenerate BB, any call to `BB::getTerminator()` will return nullptr, and at latest the IR-verifier will crash. Any `llvm_unreachable` and `assert` in OpenMPIRBuilder and elsewhere will also still cause a crash, not `mlir::emitError` which seems inconsequential.
> 
> However, of course I will not stand in the way if this is the community-chosen solution and/or established pattern. Also, if there is use case with a recoverable error we would need this as well.

I agree that once one of these errors were triggered, we'd have to stop any further translation or codegen process. But the approach I'm proposing actually does this. There is indeed some level of collaboration needed at each stage for that to happen, but basically the pattern is already to return `LogicalResult` from translation functions, always check its return value and return a failure straight away if it didn't succeed. So, once an OMPIRBuilder function fails, we can emit an error and return failure. This is checked by the caller, the caller of the caller, etc. until the original `mlir::translateModuleToLLVMIR()` call returns a null pointer instead of a half-created invalid IR.

> I might have overblown the problem. If with every `Error`/`Expected` we immediately return from the entire call stack, until there is no use of the LLVM-IR, there should not be a lot of things that could go wrong. Clang's use of OpenMPIRBuilder still needs to crash on an erroneous OpenMPIRBuilder call because it doesn't have that `Error`/`Expected` design.
> 
> MLIR uses Rust-style error propagation, and since OpenMPIRBuilder is used within MLIR, it should be compatible with it. I am still not convinced that it provides any benefit, but consider my objections retracted.

In a way, the changes to the OMPIRBuilder are not really mandatory to make use of. Callbacks passed to it from Clang can decide to exit the program on irrecoverable failure and always return success, and assume the returned `Expected` values never hold an `Error`. In MLIR this lets us integrate with the existing error reporting pattern, but it also creates the infrastructure to deal with recoverable errors in both Clang and Flang. I see it as a bit more flexible solution.

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 1]
```diff
@@ -2331,8 +2332,10 @@ void CGOpenMPRuntime::emitBarrierCall(CodeGenFunction &CGF, SourceLocation Loc,
   auto *OMPRegionInfo =
       dyn_cast_or_null<CGOpenMPRegionInfo>(CGF.CapturedStmtInfo);
   if (CGF.CGM.getLangOpts().OpenMPIRBuilder) {
-    CGF.Builder.restoreIP(OMPBuilder.createBarrier(
-        CGF.Builder, Kind, ForceSimpleCall, EmitChecks));
+    auto Result = OMPBuilder.createBarrier(CGF.Builder, Kind, ForceSimpleCall,
```
Comment (bhandarkar-pranav) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:2335]: _nit#1_: I have been receiving feedback on some of my PRs to be a little more conservative with the use of `auto`. I have concluded that it is best to use it only when it is obvious from the RHS of the assignment (e.g `dyn_cast<SomeType>`. [(Relevant part of the style guide)](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable)

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 2]
```diff
@@ -5928,8 +5931,10 @@ void CGOpenMPRuntime::emitTargetOutlinedFunctionHelper(
         return CGF.GenerateOpenMPCapturedStmtFunction(CS, D.getBeginLoc());
       };
 
-  OMPBuilder.emitTargetRegionFunction(EntryInfo, GenerateOutlinedFunction,
-                                      IsOffloadEntry, OutlinedFn, OutlinedFnID);
+  llvm::Error Err = OMPBuilder.emitTargetRegionFunction(
```
Comment (bhandarkar-pranav) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:5934]: _nit#2_: There is now some inconsistency in the use of `auto` type deduction. Here it isn't used but used previously in seemingly the same circumstances.

**Review by bhandarkar-pranav — COMMENTED** [review]
Thank you for this PR, @skatrak, a very useful change. I have minor nits but other than that, LGTM.

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp (hunk 1 above)
Comment (skatrak) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:2335]: The reason I've tended towards `auto` in this patch is that the full type in this case is `llvm::Expected<llvm::OpenMPIRBuilder::InsertPointTy>`, which is quite verbose, so I thought it was justified in this case for readability. But you're right, it doesn't look like the style guide is making an exception for such cases, so I'll look into updating the PR based on this comment.

Maybe adding an `InsertPointOrErrorTy` typedef would be a good way to address this.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 3]
```diff
@@ -1171,12 +1179,15 @@ void OpenMPIRBuilder::emitCancelationCheckImpl(Value *CancelFlag,
   // post finalization block that is known to the FiniCB callback.
   Builder.SetInsertPoint(CancellationBlock);
   if (ExitCB)
-    ExitCB(Builder.saveIP());
+    if (Error Err = ExitCB(Builder.saveIP()))
+      return std::move(Err);
```
Comment (tblah) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1183]: nit: the compiler should automatically move from a returned value whether you specify `std::move` or not (guaranteed from C++17 onwards). In general it is unhelpful to add the `std::move` here because it can prevent the return from being optimized out entirely.
```suggestion
      return Err;
```

See https://stackoverflow.com/questions/17473753/c11-return-value-optimization-or-move and https://stackoverflow.com/questions/14856344/when-should-stdmove-be-used-on-a-function-return-value

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 4]
```diff
@@ -1496,7 +1507,8 @@ IRBuilder<>::InsertPoint OpenMPIRBuilder::createParallel(
   // Let the caller create the body.
   assert(BodyGenCB && "Expected body generation callback!");
   InsertPointTy CodeGenIP(PRegBodyBB, PRegBodyBB->begin());
-  BodyGenCB(InnerAllocaIP, CodeGenIP);
+  if (Error Err = BodyGenCB(InnerAllocaIP, CodeGenIP))
+    return std::move(Err);
```
Comment (tblah) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:1511]: I'm less sure about this case because it is constructing an `Expected<>` from the `Error`. It is more like
```
return Expected<OpenMPIRBuilder::InsertPointTy>{std::move(Err)};
```

I'm not sure if this `std::move` is a guaranteed compiler optimization in the same way as the above cases.

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp (hunk 1 above)
Comment (skatrak) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp]: This is done now, thanks for the suggestion.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 3 above)
Comment (skatrak) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp]: Thank you for the pointers @tblah, I wasn't aware of this detail. I've now removed all these redundant uses of `std::move`.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 4 above)
Comment (skatrak) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp]: It looks like it works as expected in this case as well (if it created a copy it would be triggering runtime failures). I see `Expected` has an implicit move constructor for an `Error` argument, so perhaps that's what's enabling this.

**Review by tblah — APPROVED** [review]
LGTM. Thanks for the extensive cleanup

**Review by bhandarkar-pranav — APPROVED** [review]
Thank you for the changes.

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 5]
```diff
@@ -2331,8 +2332,11 @@ void CGOpenMPRuntime::emitBarrierCall(CodeGenFunction &CGF, SourceLocation Loc,
   auto *OMPRegionInfo =
       dyn_cast_or_null<CGOpenMPRegionInfo>(CGF.CapturedStmtInfo);
   if (CGF.CGM.getLangOpts().OpenMPIRBuilder) {
-    CGF.Builder.restoreIP(OMPBuilder.createBarrier(
-        CGF.Builder, Kind, ForceSimpleCall, EmitChecks));
+    llvm::OpenMPIRBuilder::InsertPointOrErrorTy AfterIP =
+        OMPBuilder.createBarrier(CGF.Builder, Kind, ForceSimpleCall,
+                                 EmitChecks);
+    assert(AfterIP && "unexpected error creating barrier");
```
Comment (Meinersbur) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:2338]: [serious] @skatrak `assert` is active by `LLVM_ENABLED_ASSERTS`, but `llvm::Error`/`llvm::Expected` uses `LLVM_ABI_BREAKING_CHECKS`. In the configuration `-DLLVM_ENABLED_ASSERTS=OFF -DLLVM_ABI_BREAKING_CHECKS=FORCE_ON` the error check is not done, but then crashes unconditional to about unchecked errors.

Comment (skatrak) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:2338]: I guess then these assert should be replaced with `if (!AfterIP) llvm_unreachable(...)`, or is there a better way to deal with this?

Comment (skatrak) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:2338]: Here's the PR to address this issue: #115863. There was indeed a cleaner solution to this.
